Adopt smart pointers in page/

rwlbuis opened the pull request:
#### a463d38eb67b255bc53844068da0d6c04aaa29b3
<pre>
Adopt smart pointers in page/
<a href="https://bugs.webkit.org/show_bug.cgi?id=279593">https://bugs.webkit.org/show_bug.cgi?id=279593</a>

Reviewed by Chris Dumez.

Adopt smart pointers in page/ based
on [alpha.webkit.UncountedCallArgsChecker] and
[alpha.webkit.NoUncountedMemberChecker] warnings.

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::sortedTrackListForMenu):
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
* Source/WebCore/page/DragController.cpp:
(WebCore::isInShadowTreeOfEnabledColorInput):
(WebCore::DragController::draggableElement const):
* Source/WebCore/page/DragController.h:
* Source/WebCore/page/SpatialNavigation.cpp:
(WebCore::rectToAbsoluteCoordinates):
* Source/WebCore/page/TextIndicator.cpp:
(WebCore::takeSnapshots):
(WebCore::initializeIndicator):

Canonical link: <a href="https://commits.webkit.org/284340@main">https://commits.webkit.org/284340@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05c3abd59db470863e890c49502873e9fffa9024

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68976 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21648 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73056 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20127 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71093 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56177 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19982 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54934 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13386 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72042 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59554 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35407 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40835 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16983 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18508 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62784 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17329 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74764 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16575 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62574 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12996 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62471 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15334 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10453 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4062 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44180 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45253 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46449 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44995 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -948,7 +948,7 @@ void LocalDOMWindow::processPostMessage(JSC::JSGlobalObject& lexicalGlobalObject
             return;
 
-        auto& vm = globalObject->vm();
+        Ref vm = globalObject->vm();
```
cc @rniwa . I think the JSC team has strong feelings about not ref'ing the VM because it's unnecessary. What's our plan?

geoffreygaren replied:
If we rename WebCore::commonVM() to WebCore::vmSingleton(), then static analysis will recognize that it doesn't need refcounting.

Then, we can change JSDOMWindowBase to:
* initialize itself using WebCore::commonVM(), instead of taking an arbitrary VM& argument (or, maybe as an initial stopgap, just RELEASE_ASSERT that they're the same);
* offer a vmSingleton() member function that returns m_vm (and maybe debug ASSERTS that they're the same);

Then, in places like this, as long as we use the right type, we can skip refcounting. And if we don't have a way to use the right type, then we don't have a compile-time proof that we can skip refcounting, so we should do the refcounting.

geoffreygaren replied:
Since this is just one rare refcount, maybe we can let it slide for now, though.

rniwa replied:
We should probably revert this change. I think it's better to fix these in en masse.

rwlbuis replied:
Indeed there are a lot of warnings about auto& vm = globalObject->vm();, a very frequent construction (even besides the generated code). Will revert that part of the PR.

rwlbuis replied:
Since the vm changes were more frequent than I remembered (I thought it was just one), I removed them all and added some new fixes for clang-tidy warnings.

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -832,47 +832,47 @@ Element* DragController::draggableElement(const LocalFrame* sourceFrame, Element
         if (m_dragSourceAction.contains(DragSourceAction::DHTML) && dragMode == UserDrag::Element) {
             state.type = DragSourceAction::DHTML;
-            return element;
+            return element.get();
```
All return statements have a RefPtr now. As a result, we can update this function to return a `RefPtr` without any ref-counting churn. The return statement would `return element;` instead of `return element.get()`.

This will actually be more efficient if the call sites of this functions need a RefPtr.

rwlbuis replied:
Nice, done.

cdumez reviewed (approved):
LGTM with suggestion.

rwlbuis committed:
Adopt smart pointers in page/
https://bugs.webkit.org/show_bug.cgi?id=279593

Reviewed by Chris Dumez.

Adopt smart pointers in page/ based
on [alpha.webkit.UncountedCallArgsChecker] and
[alpha.webkit.NoUncountedMemberChecker] warnings.

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::sortedTrackListForMenu):
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
* Source/WebCore/page/DragController.cpp:
(WebCore::isInShadowTreeOfEnabledColorInput):
(WebCore::DragController::draggableElement const):
* Source/WebCore/page/DragController.h:
* Source/WebCore/page/SpatialNavigation.cpp:
(WebCore::rectToAbsoluteCoordinates):
* Source/WebCore/page/TextIndicator.cpp:
(WebCore::takeSnapshots):
(WebCore::initializeIndicator):

Canonical link: https://commits.webkit.org/284340@main
